pkg/jobrunaggregator/jobrunbigqueryloader/uploader.go

```diff
@@ -48,46 +59,64 @@ func (o *allJobsLoaderOptions) Run(ctx context.Context) error {
 
 	logrus.Infof("Launching threads to upload test runs for %d jobs", jobCount)
 
-	waitGroup := sync.WaitGroup{}
-	errCh := make(chan error, len(jobs))
+	wg := sync.WaitGroup{}
+	jobChan := make(chan jobrunaggregatorapi.JobRow, jobCount)
+	errChan := make(chan error, jobCount)
+	// Max concurrent job imports but note there's another layer of concurrency beneath this for the file uploads
+	for i := 0; i < 20; i++ {
+		wg.Add(1)
+		go o.processJob(ctx, i, jobCount, jobChan, errChan)
+	}
+
 	for i := range jobs {
 		job := jobs[i]
+		jobChan <- job
+	}
+
+	wg.Wait()
+	close(jobChan)
+	close(errChan)
+
+	errs := []error{}
+	for e := range errChan {
+		logrus.WithError(e).Error("error encountered during upload")
+		errs = append(errs, e)
+	}
+
+	duration := time.Now().Sub(start)
+	logrus.WithFields(logrus.Fields{
+		"jobs":     jobCount,
+		"duration": duration,
+		"errors":   len(errs),
+	}).Info("completed upload")
+
+	return utilerrors.NewAggregate(errs)
+}
 
+// processJob is started in several concurrent goroutines to pull jobs to process from the jobChan. Errors are sent
+// to the errChan for aggregation in the main thread.
+func (o *allJobsLoaderOptions) processJob(ctx context.Context, workerThread, jobCount int, jobChan <-chan jobrunaggregatorapi.JobRow, errChan chan<- error) {
+	for job := range jobChan {
 		jobLogger := logrus.WithFields(logrus.Fields{
-			"job": job.JobName,
+			"job":    job.JobName,
+			"worker": workerThread,
 		})
+		// log how many job remain to be processed
+		jobLogger.WithField("remaining", fmt.Sprintf("%d/%d", len(jobChan), jobCount)).Info("pulled job from queue")
```

Comment:
this message is only correct if the for loop adding the jobs to the channel has finished before this is called.

Reply:
I've modified ordering here to populate the channel before we start the routines to process it.